### ISSUE TITLE
Trusty - update Mimio Studio software

### ISSUE THREAD

**Issue by asokero** [opinsys/puavo#2]
http://news.mimio.com/mimiostudio-software-11-download

**Comment by juhaerk** [comment]
Does not work, dies with a segmentation fault. This is probably due to a strangeness of our system, our cow/rofs root file system confuses mimio-studio for some reason, or at least that is how I would guess based on the following:

``````
```
jussieki@juer-optiplex-01:~$ gdb /var/opt/mimio/mimio-notebook
GNU gdb (Ubuntu 7.7-0ubuntu3) 7.7
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "i686-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /var/opt/mimio/mimio-notebook...(no debugging symbols found)...done.
(gdb) run
Starting program: /var/opt/mimio/mimio-notebook 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/i386-linux-gnu/libthread_db.so.1".
[New Thread 0xb4f67b40 (LWP 5672)]
[Thread 0xb4f67b40 (LWP 5672) exited]
[New Thread 0xb4f67b40 (LWP 5673)]

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xb4f67b40 (LWP 5673)]
0xb0a7562e in getrootdev_linux () from /var/opt/mimio/system/aps_componentmanager/mimiostudio.so
(gdb) backtrace
#0  0xb0a7562e in getrootdev_linux () from /var/opt/mimio/system/aps_componentmanager/mimiostudio.so
#1  0xb0a756c2 in getRootDevUUID () from /var/opt/mimio/system/aps_componentmanager/mimiostudio.so
#2  0xb0a7390f in osGetCompUUID () from /var/opt/mimio/system/aps_componentmanager/mimiostudio.so
#3  0xb0a6a904 in nalpGetCompUUID () from /var/opt/mimio/system/aps_componentmanager/mimiostudio.so
#4  0xb0a4b4ee in getComputerUUID () from /var/opt/mimio/system/aps_componentmanager/mimiostudio.so
#5  0xb0a4b5bb in getComputerID () from /var/opt/mimio/system/aps_componentmanager/mimiostudio.so
#6  0xb0a5e4f4 in shaferGetFilename () from /var/opt/mimio/system/aps_componentmanager/mimiostudio.so
#7  0xb0a5f485 in NalpLibOpen () from /var/opt/mimio/system/aps_componentmanager/mimiostudio.so
#8  0xb1294beb in ?? () from /var/opt/mimio/system/aps_componentmanager1.so
#9  0xb1294ecc in ?? () from /var/opt/mimio/system/aps_componentmanager1.so
#10 0xb129a28a in ?? () from /var/opt/mimio/system/aps_componentmanager1.so
#11 0xb54c6dd7 in ?? () from /var/opt/mimio/mimiosys.so
#12 0xb54c7520 in ?? () from /var/opt/mimio/mimiosys.so
#13 0xb5510695 in ?? () from /var/opt/mimio/mimiosys.so
#14 0xb5510943 in ?? () from /var/opt/mimio/mimiosys.so
#15 0xb5510b00 in ?? () from /var/opt/mimio/mimiosys.so
#16 0xb54ccbeb in ?? () from /var/opt/mimio/mimiosys.so
#17 0xb54cccbb in ?? () from /var/opt/mimio/mimiosys.so
#18 0xb6b3af70 in start_thread (arg=0xb4f67b40) at pthread_create.c:312
#19 0xb729170e in clone () at ../sysdeps/unix/sysv/linux/i386/clone.S:129
(gdb) kill
Kill the program being debugged? (y or n) y
(gdb) quit
```
``````

This was on trusty. Note the function name: getrootdev_linux(). We probably need upstream help on this one, because this is not open source.

I do not think mimio-studio should need to know the root device for any reason, but not seeing the code I can not tell.

**Comment by juhaerk** [comment]
I tried to strace and ltrace this, but after looking that for a while I must say I do not know what it really does (so I could somehow work around this issue).

Upstream should be contacted on this, maybe they would also fix their installer some day.

**Comment by juhaerk** [comment]
Not giving up yet! This is the disassembled code for getrootdev_linux():

```
000555f0 <getrootdev_linux>:
   555f0:       55                      push   %ebp
   555f1:       89 e5                   mov    %esp,%ebp
   555f3:       83 ec 18                sub    $0x18,%esp
   555f6:       89 5d f4                mov    %ebx,-0xc(%ebp)
   555f9:       e8 69 32 fd ff          call   28867 <geteuid@plt+0xf7>
   555fe:       81 c3 ae 0d 1a 00       add    $0x1a0dae,%ebx
   55604:       89 75 f8                mov    %esi,-0x8(%ebp)
   55607:       89 7d fc                mov    %edi,-0x4(%ebp)
   5560a:       8d 83 07 73 fd ff       lea    -0x28cf9(%ebx),%eax
   55610:       89 04 24                mov    %eax,(%esp)
   55613:       e8 98 1c fd ff          call   272b0 <getPhysDev_linux@plt>
   55618:       85 c0                   test   %eax,%eax
   5561a:       89 45 f0                mov    %eax,-0x10(%ebp)
   5561d:       74 5a                   je     55679 <getrootdev_linux+0x89>
   5561f:       8b 7d f0                mov    -0x10(%ebp),%edi
   55622:       b9 05 00 00 00          mov    $0x5,%ecx
   55627:       fc                      cld    
   55628:       8d b3 0d 73 fd ff       lea    -0x28cf3(%ebx),%esi
   5562e:       f3 a6                   repz cmpsb %es:(%edi),%ds:(%esi)
   55630:       75 5a                   jne    5568c <getrootdev_linux+0x9c>
   55632:       8b 45 f0                mov    -0x10(%ebp),%eax
   55635:       89 04 24                mov    %eax,(%esp)
   55638:       e8 f3 15 fd ff          call   26c30 <strlen@plt>
   5563d:       83 e8 04                sub    $0x4,%eax
   55640:       89 04 24                mov    %eax,(%esp)
   55643:       e8 b8 1f fd ff          call   27600 <malloc@plt>
   55648:       8b 55 08                mov    0x8(%ebp),%edx
   5564b:       89 02                   mov    %eax,(%edx)
   5564d:       8b 55 f0                mov    -0x10(%ebp),%edx
   55650:       89 04 24                mov    %eax,(%esp)
   55653:       83 c2 05                add    $0x5,%edx
   55656:       89 54 24 04             mov    %edx,0x4(%esp)
   5565a:       e8 31 19 fd ff          call   26f90 <strcpy@plt>
   5565f:       8b 45 f0                mov    -0x10(%ebp),%eax
   55662:       89 04 24                mov    %eax,(%esp)
   55665:       e8 c6 0e fd ff          call   26530 <free@plt>
   5566a:       31 c0                   xor    %eax,%eax
   5566c:       8b 5d f4                mov    -0xc(%ebp),%ebx
   5566f:       8b 75 f8                mov    -0x8(%ebp),%esi
   55672:       8b 7d fc                mov    -0x4(%ebp),%edi
   55675:       89 ec                   mov    %ebp,%esp
   55677:       5d                      pop    %ebp
   55678:       c3                      ret    
   55679:       8d 83 b2 90 fd ff       lea    -0x26f4e(%ebx),%eax
   5567f:       89 04 24                mov    %eax,(%esp)
   55682:       e8 29 1c fd ff          call   272b0 <getPhysDev_linux@plt>
   55687:       89 45 f0                mov    %eax,-0x10(%ebp)
   5568a:       eb 93                   jmp    5561f <getrootdev_linux+0x2f>
   5568c:       8b 45 f0                mov    -0x10(%ebp),%eax
   5568f:       89 04 24                mov    %eax,(%esp)
   55692:       e8 99 0e fd ff          call   26530 <free@plt>
   55697:       b8 ff ff ff ff          mov    $0xffffffff,%eax
   5569c:       eb ce                   jmp    5566c <getrootdev_linux+0x7c>
   5569e:       66 90                   xchg   %ax,%ax
```

**Comment by juhaerk** [comment]
I could work around it by lying to mimio-studio through gdb-hacks:

```
jussieki@juer-optiplex-01:~$ gdb /var/opt/mimio/mimio-mimiosys 
GNU gdb (Ubuntu 7.7-0ubuntu3) 7.7
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "i686-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /var/opt/mimio/mimio-mimiosys...(no debugging symbols found)...done.
(gdb) b getrootdev_linux
Function "getrootdev_linux" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y

Breakpoint 1 (getrootdev_linux) pending.
(gdb) run
Starting program: /var/opt/mimio/mimio-mimiosys 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/i386-linux-gnu/libthread_db.so.1".
[New Thread 0xb643ab40 (LWP 9902)]
[New Thread 0xb5674b40 (LWP 9903)]
[New Thread 0xb144bb40 (LWP 9904)]
[Thread 0xb144bb40 (LWP 9904) exited]
[New Thread 0xb144bb40 (LWP 9905)]
[Switching to Thread 0xb144bb40 (LWP 9905)]

Breakpoint 1, 0xaed5b5f6 in getrootdev_linux ()
   from /var/opt/mimio/system/aps_componentmanager/mimiostudio.so
(gdb) return (char *) "/nonexistent"
Make selected stack frame return now? (y or n) y
#0  0xaed5b6c2 in getRootDevUUID ()
   from /var/opt/mimio/system/aps_componentmanager/mimiostudio.so
(gdb) cont
Continuing.
warning: Temporarily disabling breakpoints for unloaded shared library "/var/opt/mimio/system/aps_componentmanager/mimiostudio.so"
[New Thread 0xaecffb40 (LWP 9910)]
[New Thread 0xae2ffb40 (LWP 9911)]

...
```

This should probably done to all mimio-\* binaries. Then the software appears to work okay.

NOTE THAT THIS IS NOT A SOLUTION! One could play these tricks with expect + gdb, but that is very ugly.

**Comment by juhaerk** [comment]
One can also work around this with and LD_PRELOAD-hack:

```
$ cat <<'EOF' > getrootdev_linux.c
char * getrootdev_linux(void) {
        return "/nonexistent";
}
EOF

$ cc -shared -o getrootdev_linux.so getrootdev_linux.c
$ export LD_PRELOAD=$(pwd)/getrootdev_linux.so 

$ /var/opt/mimio/mimio-mimiosys &
$ /var/opt/mimio/mimio-notebook
```

Perhaps it should return something else than "/nonexistent", but at least it does not crash....

**Comment by JuanCalle** [comment]
HI, can i get your Ubuntu installer link, we made a Special version MimioStudio 11.22 for Ubuntu 12.04. it works fine.

**Comment by juhaerk** [comment]
Okay, we'll test the later version. The package we are using is mimio-studio_11.11-0intl_i386.deb (with md5sum 48014c888377e076053522c1fe96670b), and we are testing on Ubuntu Trusty (14.04).

**Comment by JuanCalle** [comment]
I have no problem using mimio-studio_11.22-0intl_i386.deb  on Ubuntu Trusty (14.04) 64 bits.

Date: Wed, 27 Aug 2014 00:58:10 -0700

**Comment by juhaerk** [comment]
I managed to download mimio-studio_11.22-0intl_i386.deb, and it does not work for us on Ubuntu 14.04 (32-bit). The installer dies with a segmentation fault.

I do believe it does work in an ordinary Ubuntu installation. However, on our systems we do not boot directly from disk, we boot from network or mount a disk image from a filesystem as root device. Our /etc/fstab does not contain anything but /proc (indicentally, I suspect even that entry is not needed).

We have some very strong (and sound) technical reasons for managing our systems in this way, thus there is no way we are going to change our systems to ordinary Ubuntu installations only for Mimio that is almost the only problematic case for us.

However, we might be able to provide what Mimio wants, only if we could know why it breaks. Why does it break? Why does it want to know the root device, and as the mechanism for discovering that is broken in the general case, why does it fail with segmentation fault? What does it do to discover the root device (as apparently happens)?

Another problem with Mimio is that the debian package contains the installer that does the actual installation. You really should not do it that way. The installer should be used to prepare the debian package that should only contain the necessary files that are needed to run Mimio. Right now package installs "just fine" even in case the installer dies with a segmention fault. Apparently the installer requires a desktop environment running, that should not be necessary. Another problem with the installer is that it leaves 630Mb of installation files in /var/opt/mimio/tmp that are apparently not needed for running the software.

We may be able to get Mimio working without your help, but some help might be necessary for that.

**Comment by sknopp** [comment]
Hi, any solution here? I'm also using Ubuntu 14.04 LTSP for client computers (network booting).
After a long time of research and trial and error I found this thread which helped a bit during the installation. http://askubuntu.com/questions/288821/how-do-i-resolve-a-cannot-open-shared-object-file-libudev-so-0-error
Mimio-notebook starts now on clients, but you can't go into settings for example. Additionally mimio-tools does not even start but only logs "Segmentation fault" to console.
Has anyone maybe already contacted Dymo for this?

**Comment by juhaerk** [comment]
sknopp: my understanding is that JuanCalle represents Dymo. We have also been in contact by email, that is how he has found this bug report.

We have not looked at this issue any further, and (for the marginal need that we have) are staying with the 9.12 release. Unfortunately it does not have support for some of the latest hardware, that we are not buying.

This issue would probably be easy to solve if mimio was open source...

**Comment by JuanCalle** [comment]
Hi Guys, I'm user of MimioStudio on Ubunutu 14 and Fedora 20, 
MimioStudio has changed to operate in each version of Linux, for example after MimioStudio 11 only can be installed in Ubunutu 13 onwards. as I work my installations.
If Mimio Studio is already installed.
1. Eliminate all processes running with mimio-studio 
   a.from the system monitor "mimiosys"  or
   b. command Line $ sudo ps aux | grep 'mimio-studio'
2. in terminal: $ sudo apt-get remove mimio-studio
3. $ sudo apt-get purge mimio-studio
4. $ sudo apt-get autoremove

Download MimioStudio with firefox 
http://www.mimio.com/en-NA/Support/Downloads/Linux.aspx
1. look for Mimio-studio_11.30-int_i386.deb¨ under downloads folder
2. double click on the file then a follow the instruction
must wait until the installation process ends
after the installation process terminates the session must exit and re-enter to view the applications on the dashboard.
Thanks 
Juan Calle

**Comment by juhaerk** [comment]
Thank you JuanCalle! It appears that the most recent version (11.30) that we tested has fixes for the most important issues we have had with Mimio. It installs into chroot directory with apt without a fatal error, and the program does not segfault like it used to. We have not yet found any serious issues with it.

**Comment by Francesconm19** [comment]
what is it the "key mimio studio"?
